app/views/notes/show.html.erb

```diff
@@ -1,25 +1,29 @@
-  <% if @node.has_power_tag('alert') && feature("alert-#{@node.power_tag('alert')}") != "" %>
+  <% if @preview %><div class="alert alert-info"><%= raw translation('notes.show.preview') %></div><% end %>
+  <% if current_user && @node.tags.size == 0 && !@preview %><div class="alert alert-warning"><%= raw translation('notes.show.note_no_tags', url: 'javascript: document.getElementById("tags-open").click();') %></div><% end %>  
+  <% if @node.has_power_tag('status:candidate', @preview_tags) %><div class="alert alert-info">This was marked as a candidate activity -- <a href="https://publiclab.org/notes/warren/09-17-2016/what-makes-a-good-activity" >read about how to adapt its content</a> into a fully-fledged, step-by-step activity that can be easily replicated.</div><% end %>
+  <% if @node.has_power_tag('replication', @preview_tags) %><div class="alert alert-info"><%= raw translation('notes.show.replication') %> <a href="/n/<% if !@preview %><%= @node.power_tag('replication') %><% end %>"><%= raw translation('notes.show.replication_link') %></a>.</div><% end %>
+  <% if @node.has_power_tag('series', @preview_tags) %><div class="alert alert-info"><%= raw translation('notes.show.series') %> <a href="/tag/series:<% if !@preview %><%= @node.power_tag('series')%>"><%= @node.power_tag('series') %><% end %></a>.</div><% end %>
+  <% if @node.has_power_tag('build', @preview_tags) %><div class="alert alert-info"><%= raw translation('notes.show.build') %> <a href="/n/<% if !@preview %><%= @node.power_tag('build') %><% end %>"><%= raw translation('notes.show.build_link') %></a>.</div><% end %>
+
+  <% if @node.has_power_tag('alert', @preview_tags) && feature("alert-#{@node.power_tag('alert')}") != "" && !@preview %>
     <%= feature("alert-#{@node.power_tag('alert')}") %>
   <% end %>
-  <% if @node.has_power_tag('with') && current_user && @node.coauthors.collect(&:username).include?(current_user.username) %> <div class="alert alert-success"> You are a coauthor of this post. This attributes your contribution, and gives you edit access to the note. <a href="https://publiclab.org/wiki/power-tags#Coauthorship"> Learn more </a></div><% end %>
-  <% if @node.has_power_tag('upgrade') %><div class="alert alert-success"> This is an upgrade for <a href="/<%= @node.power_tag('upgrade') %>"><%= @node.power_tag('upgrade') %></a>. Try building it and <a href="/n/<%= @node.power_tag('upgrade') %>"> report back </a>how it goes.</div><% end %>
+  <% if @node.has_power_tag('with', @preview_tags)  && current_user && @node&.coauthors&.collect(&:username)&.include?(current_user.username) && !@preview  %> <div class="alert alert-success"> You are a coauthor of this post. This attributes your contribution, and gives you edit access to the note. <a href="https://publiclab.org/wiki/power-tags#Coauthorship"> Learn more </a></div><% end %>
```

Comment:
I see now, the difficulty here. Pretty tough! Was it really hard to try to create temporary `Tag` records too, without saving? I guess, if we don't save them, then the `has_power_tag` functions won't work at all.

So what we're doing is... providing a set of tags to search through /instead/ of the database, right? In a general sense. 

Could we generalize this idea then, to offer as a 2nd parameter, not "preview tags" specifically, but maybe... an `override_tags` parameter, which "overrides" the tags which would be searched? That way, `@node.has_power_tag(tag, override_tags)` would be a little easier to understand, without having to know the specifics of this kind of odd use case? What do you think of that?

We could even add a unit test that demonstrates that `@node.has_power_tag('nonexistent-tag')` returns false, but `@node.has_power_tag('nonexistent-tag', ['nonexistent-tag'])` passes?

Reply:
I guess this is just a terminology change to the `has_power_tag` modifications above. So the parameter would be more descriptive as `override_tags` and the name would indicate an easier to understand, more generalized usage. 

The unit test I mention could go here: 

https://github.com/publiclab/plots2/blob/7c365d921f7dcdfe102ba11857569eed70b09a9d/test/unit/node_test.rb#L333-L336

I think it's a relatively simple one to add and would help protect these otherwise kind of obscure changes to the node.rb methods. What do you think!

Reply:
Hi @jywarren so when I tested this locally, it only seemed to display notices related to the power tags method, I don't think it does this 

> `So what we're doing is... providing a set of tags to search through /instead/ of the database, right? In a general sense.`

At some point, I had an internal debate as to whether adding the preview tags within the power methods is really necessary because some of the notices displayed looked like what the author of the note would see and not what a user would see which is the purpose of a preview...

Creating a temporary Tag seems like a great idea.. let me do that first and see if it makes the current implementation less complicated.

Reply:
OK but i think the issue with temporary tag records is that `has_tag` and `has_powertag` both make new database queries, so the temp tags wouldn't appear in such queries. That's why i think your idea of passing a "replacement" or "override" collection of tags as an optional 2nd parameter is a great idea, now that i understood the scope of the problem. Would you like to just rename it to that more general name and we can give that a try? Let me take a look and make a suggestion, but we could confirm it with a couple simple unit tests as well.

Reply:
Yes, an override is a good idea however, it does not help in displaying that the preview post `is part of` certain tags.. like in a normally posted note...since the `_tagging.html.erb` template still hits the DB(`normal_tags`)
```
<% if @node %>
  <!-- listing of posts in response to this one -->
  <%= render partial: 'tag/replication' %>
  <!-- This is the sidebar tagging display, also renders Subscribe button for multiple subscription -->
  <p style="font-size:0.9em;color:#666;margin-top:5px;">
    <% if @node.normal_tags.count > 0 %>
      This is part of:
    <% end %>
  </p>
  <div id="tags" class="node-tags">
    <%= render partial: 'tag/tags', locals: { power_tag: false, badge_name: 'badge-secondary', tags: @node.normal_tags(:followers).includes(:tag) } %>
    <%= render partial: 'tag/tags', locals: { power_tag: true, badge_name: 'badge-primary', tags: @node.power_tag_objects.includes(:tag) } %>
  </div>
<% else %>
  <div id="tags" class="profile-tags">
    <%= render partial: 'tag/tags', locals: { power_tag: false, badge_name: 'badge-primary', tags: user.user_tags, user: user } %>
  </div>
<% end %>
```
which is another part of displaying the preview tags... 

Currently, even with the override, the code will not display related tags in the preview section...which is another issue to look at...
![Screenshot 2020-11-09 at 19 22 59](https://user-images.githubusercontent.com/7622875/98567462-ff773600-22c0-11eb-9aa7-98a3207a00ff.png)